src/rust/engine/process_execution/src/nailgun/nailgun_pool.rs

```diff
@@ -0,0 +1,315 @@
+}
+
+fn read_port(child: &mut std::process::Child) -> Result<Port, String> {
+  let stdout = child
+    .stdout
+    .as_mut()
+    .ok_or_else(|| "No Stdout found!".to_string());
+  stdout.and_then(|stdout| {
+    let reader = io::BufReader::new(stdout);
+    let line = reader
+      .lines()
+      .next()
+      .ok_or("There is no line ready in the child's output")?
+      .map_err(|err| format!("{}", err))?;
+    let port = &NAILGUN_PORT_REGEX
+      .captures_iter(&line)
+      .next()
+      .ok_or("Output for nailgun server didn't match the regex!")?[1];
+    port
+      .parse::<Port>()
+      .map_err(|e| format!("Error parsing port {}! {}", &port, e))
+  })
+}
+
+impl NailgunProcess {
+  fn start_new(
+    name: NailgunProcessName,
+    startup_options: ExecuteProcessRequest,
+    workdir_path: &PathBuf,
+    nailgun_server_fingerprint: NailgunProcessFingerprint,
+    build_id: String,
+  ) -> Result<NailgunProcess, String> {
+    let cmd = startup_options.argv[0].clone();
+    // TODO: This is an expensive operation, and thus we info! it.
+    //       If it becomes annoying, we can downgrade the logging to just debug!
+    info!(
+      "Starting new nailgun server with cmd: {:?}, args {:?}, in cwd {:?}",
+      cmd,
+      &startup_options.argv[1..],
+      &workdir_path
+    );
+    let handle = std::process::Command::new(&cmd)
+      .args(&startup_options.argv[1..])
+      .stdout(Stdio::piped())
+      .stderr(Stdio::piped())
+      .current_dir(&workdir_path)
+      .spawn();
+    handle
+      .map_err(|e| {
+        format!(
+          "Failed to create child handle with cmd: {} options {:#?}: {}",
+          &cmd, &startup_options, e
+        )
+      })
+      .and_then(|mut child| {
+        let port = read_port(&mut child);
+        port.map(|port| (child, port))
+      })
+      .and_then(|(child, port)| {
+        debug!(
+          "Created nailgun server process with pid {} and port {}",
+          child.id(),
+          port
+        );
+        Ok(NailgunProcess {
+          port: port,
+          fingerprint: nailgun_server_fingerprint,
+          name: name,
+          handle: Arc::new(Mutex::new(child)),
+          build_id: build_id,
+        })
+      })
+  }
+}
+
+impl Drop for NailgunProcess {
+  fn drop(&mut self) {
+    debug!("Exiting nailgun server process {:?}", self);
+    let _ = self.handle.lock().kill();
+  }
+}
+
+/// The fingerprint of an nailgun server process.
+///
+/// This is calculated by hashing together:
+///   - The jvm options and classpath used to create the server
+///   - The path to the jdk
+///   - The output of calling `<jdk>/bin/java -version`
+#[derive(Clone, Hash, PartialEq, Eq, Debug)]
+pub struct NailgunProcessFingerprint(pub Fingerprint);
+
+impl NailgunProcessFingerprint {
+  pub fn new(nailgun_server_req_digest: Digest, jdk_path: PathBuf) -> Result<Self, String> {
+    let jdk_version_output = std::process::Command::new(jdk_path.join("bin").join("java"))
```

Comment:
This takes about 120ms on my laptop, which is quite a lot - let's maybe just use the `realpath` of the `java` process?